### PR TITLE
task 7 done

### PR DESCRIPTION
**Pull request by crypticalfish86**
Task 7 done

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -157,3 +157,152 @@ describe('GET /api/articles', () =>
...
+describe('POST /api/articles/:article_id/comments', () =>
+{
+    test('takes an object containing a username and a body in the request body and returns an object', () =>
+    {
+        let userCommentReference = 
+            {
+                username: 'icellusedkars',
+                body: 'Fruit pastilles'
+            }
+        return request(app).post('/api/articles/1/comments').expect(200).send(userCommentReference)
```
Comment (aloush-dev) [inline __tests__/app.test.js]: post request should expect a 201

File: __tests__/app.test.js [hunk 2]
```diff
@@ -157,3 +157,152 @@ describe('GET /api/articles', () =>
...
+    })
+    test(`takes an object containing a username and a body in the request body and returns an object that has the properties:
+    body, votes, author, article_id, created_at`, () =>
+    {
+        let userCommentReference = 
+            {
+                username: 'icellusedkars',
+                body: 'Fruit pastilles'
+            }
+        return request(app).post('/api/articles/1/comments').expect(200).send(userCommentReference)
```
Comment (aloush-dev) [inline __tests__/app.test.js]: 201

File: __tests__/app.test.js [hunk 3]
```diff
@@ -157,3 +157,152 @@ describe('GET /api/articles', () =>
...
+
+
+
+
+
+describe('POST /api/articles/:article_id/comments', () =>
+{
+    test('takes an object containing a username and a body in the request body and returns an object', () =>
+    {
+        let userCommentReference =
```
Comment (aloush-dev) [inline __tests__/app.test.js]: these should be const's

File: __tests__/app.test.js [hunk 4]
```diff
@@ -157,3 +157,152 @@ describe('GET /api/articles', () =>
...
+            username: 'THIS_USER_DOES_NOT_EXIST',
+            body: 'Fruit pastilles'
+        }
+        return request(app).post('/api/articles/1/comments').expect(404).send(userCommentReference)
+        .then((err) =>
+        {
+            expect(err.body.msg).toEqual('invalid username: user not found')
+        })
+    })
+    test('if the username is matched but the reference body is not matched anywhere, will return an error 404', () =>
```
Comment (aloush-dev) [inline __tests__/app.test.js]: The body won't exist. This is supposed to be a post request where you INSERT INTO the comments database

File: app/app.js [hunk 5]
```diff
@@ -1,13 +1,23 @@
...
 
 app.get('/api/articles', getArticles);
 
 app.get('/api/articles/:article_id', getArticleById)
 
+
+
+
+
+app.post('/api/articles/:article_id/comments', getUserComment)
```
Comment (aloush-dev) [inline app/app.js]: would postUserComment not be a better name for this?

File: app/controller.js [hunk 6]
```diff
@@ -30,5 +30,33 @@ const getArticleById = (request, response, next) =>
...
+
+
+
+const getUserComment = (request, response, next) =>
+{
+    const { params, body } = request
+    fetchUserComment(params.article_id, body)
+    .then((rows) =>
+    {
+        response.status(201).send(rows)
```
Comment (aloush-dev) [inline app/controller.js]: this should be a 201

File: __tests__/app.test.js (hunk 4 above)
Comment (crypticalfish86) [reply __tests__/app.test.js]: i thought task 7 we are trying to get the relevant comment using an object containing a body for a username and body, not actually implement a comment?

File: app/controller.js [hunk 7]
```diff
@@ -30,5 +30,33 @@ const getArticleById = (request, response, next) =>
...
+
+
+
+
+
+
+const getUserComment = (request, response, next) =>
+{
+    const { params, body } = request
+    fetchUserComment(params.article_id, body)
```
Comment (aloush-dev) [inline app/controller.js]: You're calling this get and fetch when this ticket is about posting a comment.

File: app/model.js [hunk 8]
```diff
@@ -71,4 +71,127 @@ const fetchArticleById = (params) =>
...
+
+
+
+
+
+
+
+const fetchUserComment = (params, requestBody) =>
+{
+    if(!/[0-9]+/g.test(params))
```
Comment (aloush-dev) [inline app/model.js:129]: regex isn't needed here PSQL will throw an error if the Article ID is not a number

File: app/model.js [hunk 9]
```diff
@@ -71,4 +71,127 @@ const fetchArticleById = (params) =>
...
+
+
+
+
+
+
+
+
+
+const fetchUserComment = (params, requestBody) =>
```
Comment (aloush-dev) [inline app/model.js]: see previous comments about this being a post and not a fetch

**Review by aloush-dev — CHANGES_REQUESTED** [review]
You need to go back and check the code here.
You're not INSERTING into the database anywhere here and just selecting.

Have a look at your tests again and re write this model to do what the ticket is asking.

File: app/model.js [hunk 10]
```diff
@@ -119,6 +119,50 @@ const fetchArticleById = (params) =>
...
+const addUserComment = (params, body) =>
+{
+    if(!/[0-9]+/g.test(params))
+    {
+        return Promise.reject({status: 400, msg: 'invalid article ID: not a number'})
+    }
+    return db.query
+    (
+        `
+        SELECT *
```
Comment (two-triangles) [inline app/model.js:140]: params should be article_id - this is confusing if you don't know what has been passed from the controller.

File: app/model.js [hunk 11]
```diff
@@ -119,6 +119,50 @@ const fetchArticleById = (params) =>
...
+            .then(() =>
+            {
+                return db.query
+                (
+                    `
+                    SELECT *
+                    FROM comments
+                    WHERE body = \'${body.body}\'
+                    AND author = \'${body.username}\'
+                    AND article_id = ${params}
```
Comment (two-triangles) [inline app/model.js]: You are directly inserting values into your SQL query. This leads us open to the risk of SQL injection. We need to be using parametrized queries here and let postgres's server handle the substitution of these values. We have covered this in lectures over the last two weeks and it is very important that you change this.

Comment (crypticalfish86) [reply app/model.js]: am i not protected from SQL attack on the code on line 124?

File: __tests__/app.test.js [hunk 12]
```diff
@@ -158,6 +158,90 @@ describe('GET /api/articles', () =>
...
+            username: 'icellusedkars',
+            body: 'new comment test'
+        }
+        return request(app).post('/api/articles/12098376/comments').expect(404).send(userCommentReference)
+        .then((err) =>
+        {
+            expect(err.body.msg).toEqual('invalid article ID: article not found')
+        })
+    })
+})
```
Comment (two-triangles) [inline __tests__/app.test.js]: this looks good Jace. Some extra tests needed: 

201 - ignores extra properties on request body
400 - missing body/username properties 
404 - username not found

**Comment by two-triangles** [comment]
Please have a look at the tests I have asked you to implement and once you have done them, make another nchelp pr - not far off being complete now.

**Comment by two-triangles** [comment]
You have some merge conflicts that need addressing before I can review your changes. If you need some assistance with this please let me know

**Pull request merged**